pkg/controller/component/synthesize_component_test.go

```diff
@@ -0,0 +1,185 @@
+					Name: "test-compdef",
+				},
+				Spec: appsv1alpha1.ComponentDefinitionSpec{
+					Configs: []appsv1alpha1.ComponentConfigSpec{
+						{
+							ComponentTemplateSpec: appsv1alpha1.ComponentTemplateSpec{
+								Name:        "app",
+								TemplateRef: "app",
+								VolumeName:  "app",
+							},
+						},
+						{
+							ComponentTemplateSpec: appsv1alpha1.ComponentTemplateSpec{
+								Name:       "external",
+								VolumeName: "external",
+							},
+						},
+					},
+				},
+			}
+			comp = &appsv1alpha1.Component{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "test-cluster-comp",
+					Labels: map[string]string{
+						constant.AppInstanceLabelKey:     "test-cluster",
+						constant.KBAppClusterUIDLabelKey: "uuid",
+					},
+					Annotations: map[string]string{
+						constant.KubeBlocksGenerationKey: "1",
+					},
+				},
+				Spec: appsv1alpha1.ComponentSpec{
+					Configs: []appsv1alpha1.ClusterComponentConfig{},
+				},
+			}
+		})
+
+		It("comp def", func() {
+			synthesizedComp, err := buildSynthesizedComponent(reqCtx, cli, compDef, comp, nil, nil, nil, nil)
+			Expect(err).Should(BeNil())
+
+			Expect(synthesizedComp).ShouldNot(BeNil())
+			Expect(synthesizedComp.ConfigTemplates).Should(BeEquivalentTo(compDef.Spec.Configs))
+		})
+
+		It("w/ comp override - ok", func() {
+			comp.Spec.Configs = append(comp.Spec.Configs, appsv1alpha1.ClusterComponentConfig{
+				Name: func() *string { name := "external"; return &name }(),
+				ClusterComponentConfigSource: appsv1alpha1.ClusterComponentConfigSource{
+					ConfigMap: &corev1.ConfigMapVolumeSource{
+						LocalObjectReference: corev1.LocalObjectReference{
+							Name: "external-cm",
+						},
+					},
+				},
+			})
+			synthesizedComp, err := buildSynthesizedComponent(reqCtx, cli, compDef, comp, nil, nil, nil, nil)
+			Expect(err).Should(BeNil())
+
+			Expect(synthesizedComp).ShouldNot(BeNil())
+			Expect(synthesizedComp.ConfigTemplates[0]).Should(BeEquivalentTo(compDef.Spec.Configs[0]))
+
+			expectExternalConfig := compDef.Spec.Configs[1]
+			expectExternalConfig.TemplateRef = comp.Spec.Configs[0].ConfigMap.Name
+			Expect(synthesizedComp.ConfigTemplates[1]).Should(BeEquivalentTo(expectExternalConfig))
+		})
+
+		It("w/ comp override - not defined", func() {
+			comp.Spec.Configs = append(comp.Spec.Configs, appsv1alpha1.ClusterComponentConfig{
+				Name: func() *string { name := "not-defined"; return &name }(),
+				ClusterComponentConfigSource: appsv1alpha1.ClusterComponentConfigSource{
+					ConfigMap: &corev1.ConfigMapVolumeSource{
+						LocalObjectReference: corev1.LocalObjectReference{
+							Name: "external-cm",
+						},
+					},
+				},
+			})
+			_, err := buildSynthesizedComponent(reqCtx, cli, compDef, comp, nil, nil, nil, nil)
+			Expect(err).ShouldNot(BeNil())
+			Expect(err.Error()).Should(ContainSubstring("not defined in definition"))
+		})
+
+		It("w/ comp override - both specified", func() {
+			compDef.Spec.Configs[1].TemplateRef = "external"
```

Comment:
This check ensures that only the configuration entries in the ComponentDefinition with the templateRef field left empty can be set in `clusterComponentSpec.configs`. However, currently most Addons have this templateRef field specified. This could limit the usage of this feature. What is your consideration behind adding this check?

Reply:
I see. Hope in future version, items merging to the Configuration Template ConfigMap can be supported.